volatility3/framework/layers/hib.py

```diff
@@ -0,0 +1,269 @@
+# This file is Copyright 2024 Volatility Foundation and licensed under the Volatility Software License 1.0
+# which is available at https://www.volatilityfoundation.org/license/vsl-v1.0
+#
+# References:
+#   - https://www.forensicxlab.com/posts/hibernation/ : Vulgarized description of the hibernation file structure and the implementation of this layer.
+#   - https://www.cct.lsu.edu/~golden/Papers/sylvehiber.pdf : Scientific paper.
+#   - https://www.vergiliusproject.com/kernels/x64/ : Windows kernel structures used to track the evolution of the hibernation file structure in time.
+
+from typing import Optional
+import logging, struct
+from volatility3.framework import interfaces, constants, exceptions
+from volatility3.framework.layers import segmented
+from volatility3.framework.layers.codecs import (
+    lz77_plain_decompress,
+    lz77_huffman_decompress,
+)
+
+vollog = logging.getLogger(__name__)
+
+
+def uncompress(data: bytes, flag):
+    """
+    Desc:
+    Params:
+        - data: the compressed data from a compression set
+        - flag: what is the decompression algorithm to use.
+        - out_size: Size of the decompressed data
+    Return: The decompressed data (consecutive pages).
+    """
+    if flag == 0 or flag == 1:
+        return lz77_plain_decompress(data)  # See layers.codecs
+    elif flag == 2 or flag == 3:
+        return lz77_huffman_decompress(data, 65536)[0]  # See layers.codecs
+    else:
+        vollog.warning(
+            f"A compression set could not be decompressed: Compression algorithm : {flag}"
+        )
+        raise ValueError("Cannot decompress the data.")
+
+
+class HibernationLayer(segmented.NonLinearlySegmentedLayer):
+    """
+    A TranslationLayer that maps physical memory against a x64 Microsoft Windows hibernation file.
+    This Translation Layer is meant to be used in conjunction with the Hibernation.Info and Hibernation.Dump plugins.
+    """
+
+    WINDOWS_10_2016_1703_TO_23H2 = 0
+    WINDOW_8 = 1
+    WINDOWS_10_2016_1507_1511 = 2
+    WINDOWS_10_2016_1607 = 3
+
+    # TODO: Make me compatible with x86 by adding options to the Hib plugins.
+    PAGE_SIZE = 4096  # x64 page size.
+    HEADER_SIZE = 4
+    PAGE_DESC_SIZE = 8
+
+    def __init__(
+        self,
+        context: interfaces.context.ContextInterface,
+        config_path: str,
+        name: str,
+        **kwargs,
+    ):
+        """
+        Initializes the Hibernation file layer.
+        """
+        self._compressed = (
+            {}
+        )  # Keep track of which compression algorithm by each mapped compressed data.
+        self._mapping = (
+            {}
+        )  # This will hold the mapping between the PageNumber in the decompressed data vs the physical page number.
+
+        if "plugins.Dump.version" in context.config:
```

Comment:
The fact that this translation layer is meant to be used with the hibernation.Dump plugin makes me feel bad about this part.
Indeed the FirstKernelRestorePage and KernelPagesProcessed offsets can only be determine via the user input for now.

Is making a translation layer partially dependant to a plugin is an acceptable concept in the volatility3 framework ? 

This concept is working well with this implementation and seems to have no side effects from all the tests I've made but an external eye from your side will be delightful :D I propably missed something.

Reply:
If this is the only requirement for the layer, then make it a requirement (like the `kernel_virtual_offset` for Intel layers), then anything can instantiate it as long as they provide the right version number for the layer (it might be better if the layer could figure it out for itself, but if not then ask the instantiator to provide it).

Reply:
If this is the only requirement for the layer, then make it a requirement (like the kernel_virtual_offset for Intel layers), then anything can instantiate it as long as they provide the right version number for the layer (it might be better if the layer could figure it out for itself, but if not then ask the instantiator to provide it).

Reply:
Did not find a way to figure the version out by itself in my research at time of writing, I'll look at what’s done in the Intel layer thanks for the tip!

Reply:
Probably add a `PluginRequirement`[https://github.com/volatilityfoundation/volatility3/blob/develop/volatility3/framework/configuration/requirements.py#L569] and that should do you.  5:)